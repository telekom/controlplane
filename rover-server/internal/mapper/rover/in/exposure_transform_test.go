// Copyright 2025 Deutsche Telekom IT GmbH
//
// SPDX-License-Identifier: Apache-2.0

package in

import (
	"github.com/gkampitakis/go-snaps/snaps"
	. "github.com/onsi/ginkgo/v2"
	. "github.com/onsi/gomega"
	roverv1 "github.com/telekom/controlplane/rover/api/v1"

	"github.com/telekom/controlplane/rover-server/internal/api"
)

var _ = Describe("Exposure Transformation Mapper", func() {
	Context("mapExposureTransformation", func() {
		It("must map removeHeaders correctly", func() {
			// Given
			input := api.ApiExposure{
				BasePath:      "/test",
				RemoveHeaders: []string{"X-Test-Header", "Authorization"},
			}

			output := &roverv1.ApiExposure{}

			// When
			mapExposureTransformation(input, output)

			// Then
			Expect(output.Transformation).ToNot(BeNil())
			Expect(output.Transformation.Request.Headers.Remove).To(HaveLen(2))
			Expect(output.Transformation.Request.Headers.Remove).To(ContainElements("X-Test-Header", "Authorization"))
			snaps.MatchSnapshot(GinkgoT(), output.Transformation)
		})

		It("must handle nil removeHeaders", func() {
			// Given
			input := api.ApiExposure{
				BasePath: "/test",
				// RemoveHeaders is nil
			}

			output := &roverv1.ApiExposure{}

			// When
			mapExposureTransformation(input, output)

			// Then
			Expect(output.Transformation).To(BeNil())
		})

		It("must handle empty removeHeaders", func() {
			// Given
			input := api.ApiExposure{
				BasePath:      "/test",
				RemoveHeaders: []string{},
			}

			output := &roverv1.ApiExposure{}

			// When
			mapExposureTransformation(input, output)

			// Then
			// Empty removeHeaders still creates a Transformation object
			Expect(output.Transformation).ToNot(BeNil())
			Expect(output.Transformation.Request.Headers.Remove).To(BeEmpty())
		})
	})
})
