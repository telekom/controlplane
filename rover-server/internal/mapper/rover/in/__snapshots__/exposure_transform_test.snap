
[Exposure Transformation Mapper mapExposureTransformation must map removeHeaders correctly - 1]
&v1.Transformation{
    Request: v1.RequestResponseTransformation{
        Headers: v1.HeaderTransformation{
            Remove: {"X-Test-Header", "Authorization"},
            Add:    nil,
        },
    },
}
---
